paraatm/bdn/_bdn.py

```diff
@@ -0,0 +1,209 @@
+        
+        
+        For more information about dnn model, please refer to:
+        https://en.wikipedia.org/wiki/Deep_learning#Deep_neural_networks
+        
+        Returns
+        -------
+        A sequential dnn model ready to be fit in 
+        """
+        if len(self.x_train.shape)!=2:
+            raise Exception('expect a 2 dimension array!')
+        in_shape = self.x_train.shape[-1]
+        model=Sequential()
+        model.add(Dense(self.num_unit,kernel_regularizer=l2(self.weight_decay),
+                       bias_regularizer=l2(self.weight_decay),
+                        kernel_initializer='random_uniform',
+                        input_dim=in_shape))
+        
+        model.add(Activation('relu'))
+        model.add(Dropout(self.idrop))
+        model.add(Dense(self.num_unit,kernel_regularizer=l2(self.weight_decay),
+                       bias_regularizer=l2(self.weight_decay),
+                        kernel_initializer='random_uniform',
+                       ))
+        model.add(Dropout(self.idrop))
+        model.add(Activation('relu'))  
+        if self.odrop:
+            model.add(Dropout(self.odrop))
+        model.add(Dense(1,activation='linear',
+                        kernel_regularizer=l2(self.weight_decay),
+                        bias_regularizer=l2(self.weight_decay)))
+        optimizer_=Adam(self.lr)
+        model.compile(loss='mse',metrics=['mse'],optimizer=optimizer_)
+        return model        
+        
+
+    def pred(self,flag_1 = True,flag_2 = 'rnn',custom_model = None):
+        """
+        pred is used to generate results for the deep learning model
+        
+        Parameters
+        ----------
+        flag_1: Boolen
+            By default: it's set as 'True'
+            flag_1 == True: prediction with uncertainty
+            flag_1 == False: deterministic result
+        
+        flag_2: str
+            By default, it's set as 'rnn'
+            flag_2 == 'rnn': generate results for the embedded rnn model
+            flag_2 == 'dnn': generate results for the embedded dnn model
+            flag_2 == 'custom': generate results for the customized model
+            When flag_2 == 'custom', a customized model must be provided
+        
+        custom_model: None is set by default, otherwise is a  customized sequential model
+        
+        Returns
+        -------
+        An array which is predictions for the test data
+        The array has a shape as num_test_data by num_iter
+        num_test_data:  Number of test data
+        num_iter: number of iterations
+        """
+        if flag_2 =='rnn':
+            model = self._get_rnn_model()
+        if flag_2 =='dnn':
+            model = self._get_dnn_model()
+        if flag_2 =='custom':
+            model = custom_model
+        #print(model.summary())
+        model.fit(self.x_train,self.y_train,self.batch_size,
+                      self.epochs,verbose = False)
```

Comment:
Would it be better to move the `model.fit` call under `__init__` so that fitting is not repeated every time `pred` is called?  Depending on what makes the most sense, you could either take the model type flag as an input to `__init__`, or you could fit all model types in `__init__` and then set `model` to point to the appropriate structure in `pred`.

Reply:
Hi John,

Thank you for the comments and suggestion.

The model can only be fitted in after the _get_dnn_model or _get_rnn_model is called and returned an model. The neural network model I'm using right now is a sequential model, which means that it starts with input layer, then hidden layers, then output layer. After the model structure is built, it would be fit in. As the DNN model and RNN model doesn't use same data. In DNN, the input data is 2D; in RNN, the input data is 3D. So probably there is no redundancy here.  For 'you could fit all model types in __init__ and then set model to point to the appropriate structure in pred', I don't really understand it very well. Could you give me a more detailed explanation or a simple example?

Reply:
Here is the basic idea of how this could be done:

```python
class Bdn:
    def __init__(self):
        # Fit models so that they are ready to use when predictions
        # are requested
        self.rnn = self._get_rnn_model()
        self.dnn = self._get_dnn_model()
    def pred(self, x, model_type='rnn'):
        if model_type == 'rnn':
            return self.rnn.predict(x)
        elif model_type == 'dnn':
            return self.dnn.predict(x)
```

If you were concerned that you might not need both DNN and RNN to be fit, you could do something like this:
```python
class Bdn:
    def __init__(self):
        # Don't fit models until they are first needed
        self.rnn = None
        self.dnn = None
    def pred(self, x, model_type='rnn'):
        if model_type == 'rnn':
            if self.rnn is None:
                # Fit the model and store it so that it is available
                # the next time it is needed
                self.rnn = self._get_rnn_model()
            return self.rnn.predict(x)
        elif model_type == 'dnn':
            if self.dnn is None:
                self.dnn = self._get_dnn_model()
            return self.dnn.predict(x)
```

Again, I don't know the in's and out's of this particular method that you are using.  However, this is normally how data driven models would be implemented.  Fitting can be expensive, so you want to only do it once.  For example, what if the user wants to do something like this:
```python
model = Bdn(Xtrain, Ytrain)
results = model.predict(x)
# Do some calculations, then make another prediction
results = model.predict(x)
# Do some calculations, then comptue a series of predictions
for i in range(1000):
    results = model.predict(...)
```

Your code would require re-fitting the model every time the user calls `pred`, but that is unecessary.  It should be possible to fit the model once and then re-use it for predictions as many times as the user wants in the same program.